### 1. Reading the Table
Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe

In [1]:
import pandas as pd

wiki_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
read_table = pd.read_html(wiki_url,header=[0])
df = read_table[0]
df.head(10)

Postal Code           Borough                                Neighbourhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

In [2]:
df.shape

(180, 3)

### 2. Renaming Data Frame 

In [3]:
df = df.rename(columns={"Postal Code":"Postalcode", "Neighbourhood":"Neighborhood"}, index=str)
df.head()

Postalcode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

### 3. Ignore cells with a borough that is Not assigned. 
Only process the cells that have an assigned borough. 

In [4]:
df = df[df.Borough != 'Not assigned']
df.reset_index(drop=True, inplace=True)
df.head()

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### 4. Grouping neighborhood based on same Postcalcode


In [5]:
df = df.groupby("Postalcode", as_index=False).agg(lambda x: ','.join(set(x.dropna())))
df.head()

Postalcode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [6]:
df.shape

(103, 3)

### 5. If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough

In [7]:
df.loc[df['Neighborhood'] == 'Not assigned','Neighborhood'] = df['Borough']
df.head()

Postalcode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

### 6. Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making

I put markdown cells for every task I did. From this work, we can see that the shape of real dataset is (180, 3). 
In step 3, we ignore cells with a borough that is not assigned. This step made the shape of the dataset become (103, 3).
In step 5, if a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. After we did this step, the shape of the dataset did not change. It means that there is no cell that has a Not assigned neighborhood

### 7. In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [8]:
df.shape

(103, 3)